src/compile/mark/valueref.ts

```diff
@@ -135,11 +132,21 @@ export function midPoint(channel: Channel, channelDef: ChannelDef<string>, scale
             return fieldRef(channelDef, scaleName, {binSuffix: 'mid'});
           }
           // For non-stack, we can just calculate bin mid on the fly using signal.
-          return binMidSignal(channelDef, scaleName);
+          return binMidSignal(vgField(channelDef, {expr: 'datum'}),
+                              vgField(channelDef, {binSuffix: 'end', expr: 'datum'}),
+                              scaleName);
         }
         return fieldRef(channelDef, scaleName, binRequiresRange(channelDef, channel) ? {binSuffix: 'range'} : {});
       }
 
+      if (isBinned(channelDef.bin)) {
```

Comment:
Use `else if` and connect it with the clause above.

Reply:
Done